[Hardware][Intel] Generate custom activation ops using torch.compile for CPU backend.

bigPYJ1151 opened the pull request:
Generate custom activation ops using ```torch.compile``` for CPU backend.

Main changes to vLLM:
- ~~Add ```_forward_native_impl``` to each custom ops to avoid recompilation caused by tracing ```self```.~~

For vicuna-7b-v1.5, there is no significant regression:
```
# baseline
    Throughput: 1.39 requests/s, 667.72 tokens/s
# torch.compile
    Throughput: 1.38 requests/s, 662.98 tokens/s
```
For gpt-j-6b, vectorized math functions provide some improvement:
```
# baseline
    Throughput: 1.40 requests/s, 628.76 tokens/s
# torch.compile
    Throughput: 1.60 requests/s, 716.13 tokens/s
```
---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

youkaichao commented:
Hi, do you face the same problem as https://github.com/vllm-project/vllm/pull/3985 ? You are explicitly separating the `self` argument.

bigPYJ1151 commented:
@youkaichao Yes, the guard logging shows different ```self``` and ```batchsize=1``` will cause recompilation. I didn't find other more appropriate methods to hint the compiler stops the recompilation for each class instance.

youkaichao commented:
FYI: pytorch has a [flag](https://github.com/pytorch/pytorch/issues/125652) to change the behavior. and we are working with pytorch team to make it default. hopefully you don't need it in the future.

bigPYJ1151 committed:
Generate custom activation ops using torch.compile for CPU backend

bigPYJ1151 commented:
@youkaichao Thanks for your information! The flag worked. It is strange only ```RMSNorm``` is affected by this problem.
The native forward of ```RotaryEmbedding``` also uses ```cos_sin_cache``` from ```self``` but there is no recompilation.

Anyway, adding ```_forward_native_impl``` to separate ```self``` is not required for the activation ops. The changes to public interfaces are reduced.

zhouyuan commented on vllm/worker/cpu_worker.py:
```diff
@@ -177,6 +177,9 @@ def init_device(self) -> None:
 
     def load_model(self):
         self.model_runner.load_model()
+        logger.info("Warming up...")
```
is this for debug?

bigPYJ1151 replied:
just printing some status messages, because the compilation may take a while.

WoosukKwon replied:
Can we add `The compilation may take several minutes.` or something like that?

bigPYJ1151 committed:
Fix max_prompt_len for warming up inputs

WoosukKwon commented on vllm/worker/cpu_model_runner.py:
```diff
@@ -89,6 +94,24 @@ def load_model(self) -> None:
             scheduler_config=self.scheduler_config,
             cache_config=self.cache_config)
 
+        # Apply torch.compile to custom ops
+        from vllm.model_executor.custom_op import CustomOp
```
Why do we need lazy import here?

bigPYJ1151 replied:
For now, we only compile the ```CustomOps```, so import the class for type identification.

WoosukKwon replied:
Oh actually my question was why we import these "lazily".

bigPYJ1151 replied:
No special reason. I think importing related classes at the local scope will make maintenance more convenient (for example, adding more transformation or moving the procedure to other places).

WoosukKwon replied:
Thanks for the explanation. I personally think it's always good to avoid lazy imports whenever possible, but I agree that it can be a matter of personal preference. I'm ok with keeping it.

WoosukKwon commented on vllm/worker/cpu_model_runner.py:
```diff
@@ -89,6 +94,24 @@ def load_model(self) -> None:
+                                                         True,
+                                                         "cpp_wrapper": True,
+                                                         "dce": True
+                                                     })
```
Add `return` here for calrity?

WoosukKwon commented on vllm/model_executor/layers/layernorm.py:
```diff
@@ -67,6 +67,18 @@ def forward_cuda(
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
+        # Note: the forward_native() with torch.compile has significant
+        # performance regression.
+        return self.forward_cuda(
+            x,
+            residual,
+        )
```
nit:
```suggestion
        return self.forward_cuda(x, residual)
```

WoosukKwon commented on vllm/model_executor/layers/rotary_embedding.py:
```diff
@@ -232,6 +232,22 @@ def forward_tpu(
+        key: torch.Tensor,
+        offsets: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Note: the forward_native() with torch.compile has significant
+        # performance regression.
```
QQ: Have you tried `forward_native2` which was added in the TPU integration PR? For TPUs, `forward_native2` works much better than `forward_native`.

bigPYJ1151 replied:
I did a quick test but it still has significant performance regression. I also saw a warning that torchInductor doesn't support complex operations (```freqs_cis``` is complex type).

WoosukKwon commented on vllm/model_executor/layers/rotary_embedding.py:
```diff
@@ -232,6 +232,22 @@ def forward_tpu(
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Note: the forward_native() with torch.compile has significant
+        # performance regression.
+        return self.forward_cuda(
+            positions,
+            query,
+            key,
+            offsets,
+        )
```
nit:
```suggestion
        return self.forward_cuda(positions, query, key, offsets)
```

WoosukKwon commented on vllm/worker/cpu_model_runner.py:
```diff
@@ -361,3 +387,41 @@ def execute_model(
+
+        for seq_len in [1, max_num_seqs, max_num_batched_tokens]:
+            if vlm_config:
+                seq_data, dummy_multi_modal_data = MULTIMODAL_REGISTRY \
+                    .dummy_data_for_profiling(seq_len, model_config, vlm_config)
```
nit: Let's avoid using `\` and use parentheses instead.
```suggestion
                seq_data, dummy_multi_modal_data = (MULTIMODAL_REGISTRY
                    .dummy_data_for_profiling(seq_len, model_config, vlm_config))
```

WoosukKwon commented on vllm/worker/cpu_model_runner.py:
```diff
@@ -361,3 +387,41 @@ def execute_model(
         return output
+
+    @torch.inference_mode()
+    def profile_run(self) -> None:
```
What is the purpose of the profiling urn? Is the goal invoking `torch.compile` for different input shapes? Or is it for measuring CPU memory usage?

bigPYJ1151 replied:
Yes, to invoking ```torch.compile``` for ```batchsize=1``` and ```batchsize=others```

WoosukKwon replied:
Why are we profiling for those particular shapes? IIRC, `torch.compile` supports dynamic shapes unless some advanced features are used (e.g., CUDA graphs).

bigPYJ1151 replied:
Because I noticed ```torch.compile``` will generate different code for ```batchsize=1``` and ```batchsize=others``` under the dynamic mode. So we should invoke them all.

WoosukKwon replied:
Got it. Thanks for the clarification.

WoosukKwon reviewed (commented):
@bigPYJ1151 Thanks for the PR! QQ: Can we remove any C++ kernels after merging this PR?

bigPYJ1151 committed:
Fix format

bigPYJ1151 committed:
remove activation.cpp

bigPYJ1151 committed:
Format

bigPYJ1151 commented:
Hi @WoosukKwon thanks for your review! I have fixed the code style and added some notes to solve your comments, please check them. With this PR, the C++ activation functions can be removed.

WoosukKwon commented on vllm/worker/cpu_model_runner.py:
```diff
@@ -361,3 +387,46 @@ def execute_model(
+        model_config = self.model_config
+        vlm_config = self.vision_language_config
+
+        # Enable top-k sampling to reflect the accurate memory usage.
```
nit: This comment is a bit confusing since we are not profiling memory here.
```suggestion
```

bigPYJ1151 replied:
Removed

WoosukKwon commented on vllm/worker/cpu_model_runner.py:
```diff
@@ -89,6 +94,24 @@ def load_model(self) -> None:
+        # Apply torch.compile to custom ops
+        from vllm.model_executor.custom_op import CustomOp
+
+        def replace_model(model: torch.nn.Module):
+            for _, m in model.named_children():
+                if isinstance(m, CustomOp):
+                    m.forward_native = torch.compile(m.forward_native,
+                                                     dynamic=True,
+                                                     options={
+                                                         "fx_graph_cache":
+                                                         True,
+                                                         "cpp_wrapper": True,
+                                                         "dce": True
+                                                     })
```
Just wondering, can we do this in `CustomOp.forward_cpu` instead?

bigPYJ1151 replied:
I tried this, but it didn't work.
- The ```forward``` actually uses ```_forward_method```, so we should replace ```_forward_method```.
- If we replaced ```_forward_method```, the ```torch.complie``` will raise a error ```mutable rms_norm.default is not supported with cpp_wrapper```. Seems ```cpp_wrapper``` is not compatible with ```RMSNorm.forward_cuda```.

WoosukKwon replied:
I see. This doesn't look aesthetically good to me, but I don't have an alternative solution... 😞

WoosukKwon replied:
@youkaichao Could you please take a look at this part of code if you have time? Just wondering if you have any suggestion, as the code doesn't look ideal to me.

WoosukKwon reviewed (commented):
@bigPYJ1151 Thanks for updating the PR! Left more comments.

bigPYJ1151 committed:
rebase

bigPYJ1151 committed:
fix

bigPYJ1151 committed:
FIX

hmellor commented:
@bigPYJ1151 shall we close this PR as it's ben superseded by #7110?
